### ISSUE TITLE
⚡ Re-implement emit optimisations

### ISSUE THREAD

**Issue by Alpvax** [ren-lang/compiler#99]
I need to thoroughly read through the new changes before I begin, but I will attempt the following:

- [ ] Don't render IIFE when not required (only expression inside function body) (Reimplement #76)
- [ ] Don't emit double parentheses around function arguments / strip non-required parentheses (Reimplement #78, possibly also fix #87)

**Comment by andy-dot-dev** [comment]
Happy to pair or answer any questions on discord whenever. Yeah sorry there's quite a few regressions on the emit stuff I think... The way things are kinda structured at the moment is the emitters return a config object that the caller can then use to, say, wrap something in an IIFE.

An expansion on that idea could probably be good.

**Comment by Alpvax** [comment]
Yeah, I linked the old approaches so that I could compare the old approach.
The old approach for stripping double parens was to create a `singleTerm` function, which just conditionally wrapped expressions, so that should be pretty easy to implement with the current approach.
I'll also add the `esOperatorPrecedence` function as a let binding of the infix emitter, and re-implement the infix precedence parentheses. Unless you want to expand precedence across more than just the infix operators?

**Comment by hayleigh-dot-dev** [comment]
Nah that all sounds good to me!

**Comment by hayleigh-dot-dev** [comment]
Genius swapping github accounts mid-conversation...
